app/src/main/kotlin/at/bitfire/davdroid/ui/intro/BatteryOptimizationsFragment.kt

```diff
@@ -53,28 +87,42 @@ class BatteryOptimizationsFragment: Fragment() {
 
 
     override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
-        val binding = IntroBatteryOptimizationsBinding.inflate(inflater, container, false)
-        binding.lifecycleOwner = viewLifecycleOwner
-        binding.model = model
-
-        model.shouldBeWhitelisted.observe(viewLifecycleOwner) { shouldBeWhitelisted ->
-            @SuppressLint("BatteryLife")
-            if (shouldBeWhitelisted && !model.isWhitelisted.value!!)
-                ignoreBatteryOptimizationsResultLauncher.launch(BuildConfig.APPLICATION_ID)
-        }
-        binding.batteryText.text = getString(R.string.intro_battery_text, getString(R.string.app_name))
-
-        binding.autostartHeading.text = getString(R.string.intro_autostart_title, WordUtils.capitalize(Build.MANUFACTURER))
-        binding.autostartText.setText(R.string.intro_autostart_text)
-        binding.autostartMoreInfo.setOnClickListener {
-            UiUtils.launchUri(requireActivity(), App.homepageUrl(requireActivity()).buildUpon()
-                    .appendPath("faq").appendPath("synchronization-is-not-run-as-expected")
-                    .appendQueryParameter("manufacturer", Build.MANUFACTURER.lowercase(Locale.ROOT)).build())
+        return ComposeView(requireContext()).apply {
+            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
+            setContent {
+                MdcTheme {
+                    var dontShowBattery by remember { mutableStateOf(model.dontShowBattery.get()) }
```

Comment:
`dontShowBattery` is an `ObservableBoolean`, which had LiveData-equivalent functionality for XML data binding. However we shouldn't use it anymore when we use Compose.

Maybe using `MutableLiveData` would be an option? I think we could call `set` directly from the UI in this case (because the model API is that this value can be set – so it's not modifying some model-internal variable, which would be bad), and observe the value for the UI.

Reply:
I've created this wrapper for `MutableLiveData` and `SettingsManager`:
```kotlin
class PreferencesLiveData<Type : Any>(
    private val settings: SettingsManager,
    val key: String,
    private val defaultValue: Type
) : MutableLiveData<Type>() {
    private val preferenceChangeListener = SettingsManager.OnChangeListener {
        updateValue()
    }

    @Suppress("UNCHECKED_CAST")
    private fun updateValue() {
        value = when (defaultValue) {
            is Boolean -> settings.getBooleanOrNull(key) as Type? ?: defaultValue
            is Int -> settings.getIntOrNull(key) as Type? ?: defaultValue
            is Long -> settings.getLongOrNull(key) as Type? ?: defaultValue
            is String -> settings.getString(key) as Type? ?: defaultValue
            else -> throw IllegalArgumentException("Unsupported type (${defaultValue::class.java})")
        }
    }

    override fun onActive() {
        super.onActive()
        updateValue()
        settings.addOnChangeListener(preferenceChangeListener)
    }

    override fun onInactive() {
        super.onInactive()
        settings.removeOnChangeListener(preferenceChangeListener)
    }
}
```
Then in the ViewModel we just have to do:
```kotlin
val hintBatteryOptimizations = PreferencesLiveData(settings, HINT_BATTERY_OPTIMIZATIONS, true)
```

Reply:
>            else -> throw IllegalArgumentException("Unsupported type (${defaultValue::class.java})")

I like the approach, but such assumptions have to made carefully, they can crash the app (and already did in some cases). `IllegalArgumentException` should not be thrown if it can be avoided (better keep it `null` and log a warning).

Maybe we can provide a LiveData-returning `fun getBooleanLive() : LiveData` in the SettingsManager? This LiveData could maybe be an inner class and thus directly access `this@SettingsManager`. Such a Settings-LiveData could be useful in other models.

Reply:
I've added `getBooleanLive`, and made it nullable (without a default value). I think this way it's more useful, since you can actually remove the stored value, by setting it to null.